app/controllers/projects_controller.rb

```diff
@@ -105,29 +105,26 @@ def recess
 
   # PUT /projects/1/join
   def join
-    # FIXME: This is a validation
-    if @project.aasm_state == 'invention'
-      redirect_to project, error: "You can't join this project as it's finished."
-    end
-
-    if @project.join! current_user
-      message = "Welcome to the project #{current_user.name}!"
-    else
-      message = 'You already joined this project'
+    respond_to do |format|
+      if @project.join!(current_user)
+        format.js { flash['success'] = "Welcome to the project #{current_user.name}." }
```

Comment:
This could be `flash.now` instead of `flash`, so I assume that `flash.discard` will be not needed